runtime/unix.c

```diff
@@ -478,3 +481,76 @@ void caml_init_os_params(void)
   caml_sys_pagesize = sysconf(_SC_PAGESIZE);
   return;
 }
+
+void *caml_plat_mem_map(uintnat size, uintnat alignment, int reserve_only)
+{
+#ifdef __CYGWIN__
+  uintnat alloc_sz = size;
+#else
+  uintnat alloc_sz = size + alignment;
+  uintnat base, aligned_start, aligned_end;
+#endif
+  void* mem;
+
+  mem = mmap(0, alloc_sz, reserve_only ? PROT_NONE : (PROT_READ | PROT_WRITE),
+             MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+  if (mem == MAP_FAILED)
+    return 0;
+
+#ifndef __CYGWIN__
+  /* trim to an aligned region */
+  base = (uintnat)mem;
+  aligned_start = (base + alignment - 1) & ~(alignment - 1);
+  aligned_end = aligned_start + size;
+  caml_gc_message(0x1000, "munmap %" ARCH_INTNAT_PRINTF_FORMAT "d"
+                          " bytes at %p for heaps\n",
+                          aligned_start - base, (void*)base);
+  munmap(mem, aligned_start - base);
+  caml_gc_message(0x1000, "munmap %" ARCH_INTNAT_PRINTF_FORMAT "d"
+                          " bytes at %p for heaps\n",
+                          (base + alloc_sz) - aligned_end, (void*)aligned_end);
+  munmap((void*)aligned_end, (base + alloc_sz) - aligned_end);
+  mem = (void*)aligned_start;
+#endif
+
+  return mem;
+}
```

Comment:
Could you un-factor the Cygwin and the non-Cygwin versions as two separate definitions? It would be nice to explain in a comment why Cygwin does not need to care about alignment, possibly by referring to the `win32`-version comment.

Reply:
Yes - will do!